### ISSUE TITLE
Aggregation of hidden layer representations from multiple GPUs to CPU using Module/Model API

### ISSUE THREAD

**Issue by bhokaal** [apache/incubator-mxnet#4816]
Hi,
     Thanks for creating such an awesome library, it is simply the best one out in the market, in terms of speed, memory and ease of use.

I want to be able to create multiple towers with same weights parameters, like in Tensorflow, and pass  different mini-batches of data to each tower and in the end be able to concatenate the hidden layer representations to create a bigger batch on which I would like to run my loss function and propagate the gradients back. For example - 

Number of GPUs: 2 [gpu_0, gpu_1]
Network definition: 
`data = mx.symbol.Variable('data')`
`fc1 = mx.symbol.FullyConnected(data, name='fc1', num_hidden=128)`
`act1 = mx.symbol.Activation(fc1, name='relu1', act_type='relu')`
`fc2 = mx.symbol.FullyConnected(act1, name='fc2', num_hidden=64)`

Now, I want to supply a mini-batch of 100 samples such that fc2 for 50 samples (a 50 by 64 matrix) is computed on gpu_0 (lets call it fc2_0) and fc2 for the rest of the 50 samples is computed on gpu_1 (lets call it fc2_1). Lastly, I want to concatenate fc2_0 and fc_1 to get fc ( a 100 by 64 matrix) ie
` fc = mx.symbol.concat([fc1_0,fc1_1],2,0)`

**While doing so, I want various parameters of the models to be the same across the two gpus ie. fc1_weight, fc1_bias, fc2_weight and fc2_bias should be the same for both the GPUs.** 

Now, I want to apply a loss function on fc, say I have a loss function customLoss:
`loss = customLoss(fc)`

Lastly, I would like to use Module/Model API to train the network by using something like:

`model = mx.model.FeedForward.create(
         loss,
         X=data_set,
         num_epoch=num_epoch,
         learning_rate=0.01)
`
 OR 
`mod = mx.mod.Module(loss)`
`mod.fit(train_dataiter, eval_data=eval_dataiter,
            optimizer_params={'learning_rate':0.01, 'momentum': 0.9},
            num_epoch=n_epoch)`

The reason I want to aggregate the representations for loss computation is because the loss computations depends on the number of samples in the mini-batch and simply computing the loss individually on separate GPUs and averaging the gradient (the current multi-GPU of MxNet) will lead to sub-optimal performance. This is one major feature that I could not figure out in MxNet, may be it already exists but I am not sure how to achieve it. Any help would be deeply appreciated.

**Comment by piiswrong** [comment]
See the model parallel lstm example here: https://github.com/dmlc/mxnet/blob/master/example/model-parallel-lstm/lstm_ptb.py

I'm not sure if module api supports this

**Comment by bhokaal** [comment]
Thanks for the response, I will look into it. I will bug you again, if need any further help.

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
